azurelinuxagent/ga/exthandlers.py

```
@@ -1380,20 +1382,56 @@ def download(self):
 
         package_exists = False
         if os.path.exists(package_file):
-            self.logger.info("Using existing extension package: {0}", package_file)
+            msg = "Using existing extension package: {0}".format(package_file)
+            self.logger.info(msg)
+            self.report_event(message=msg)
             if self._unzip_extension_package(package_file, self.get_base_dir()):
                 package_exists = True
             else:
                 self.logger.info("The existing extension package is invalid, will ignore it.")
 
         if not package_exists:
             is_fast_track_goal_state = self.protocol.get_goal_state().extensions_goal_state.source == GoalStateSource.FastTrack
-            self.protocol.client.download_zip_package("extension package", self.pkg.uris, package_file, self.get_base_dir(), use_verify_header=is_fast_track_goal_state)
+
+            # Validate signature if extension is signed, signature validation is enabled via conf, and openssl version is supported.
+            # If validation fails, handle and report the error but do not block download; blocking will be enforced once
+            # we gain confidence in the validation process.
+            #
+            # TODO: allow users to opt-in to validation enforcement using policy, as a temporary work-around until validation is always enforced.
+            if conf.get_signature_validation_enabled() and openssl_version_supported_for_signature_validation() and self.ext_handler.encoded_signature is not None:
```

We should only validate signature in WireClient.download_zip_package() if conf flag is enabled and openssl version is supported. 

Checking these conditions in ExtHandlerInstance.download() seemed like the simplest option to me - if we moved this check to download_zip_package(), we'd probably need to raise an error or return a value to indicate if signature validation succeeded/failed/skipped. If anyone can think of a better option, let me know.

---

Makes sense for caller to check. Rather than having download_zip_package in both if/else conditions, could you just pass signature=None if these conditions aren't met?

For ex:

```
signature=None
signature_validated=False
if conf.get_signature_validation_enabled() and openssl_version_supported_for_signature_validation() and self.ext_handler.encoded_signature is not None:
    signature = self.ext_handler.encoded_signature 
try:
    self.protocol.client.download_zip_package("extension package", self.pkg.uris, package_file,
                                                              self.get_base_dir(),
                                                              use_verify_header=is_fast_track_goal_state,
                                                              signature=signature,
                                                              package_name=self.ext_handler.name,
                                                              package_version=self.ext_handler.version)
    if signature is not None:
      signature_validated=True
except SignatureValidationError as ex:
    logger.error(ustr(ex))
    add_event(op=WALAEventOperation.SignatureValidation, message=ustr(ex), name=self.ext_handler.name, version=self.ext_handler.version, is_success=False)


 
```

---

I may be missing something, but I don't see why the exthandler needs to worry about openssl_version_supported_for_signature_validation()

This is just a temporary hack for the telemetry release. Can't the signature validation code check for this condition and  just skip the validation (mark it as successful) if the openssl version is not supported?

---

My concern is that we'd skip the validation but still end up creating a state file, falsely giving the impression that signature was validated when it wasn't. That state file might be left behind for future releases, when we do implement timestamp validation and support older openssl versions. If that's not a problem (we could use a different state file for future releases), then I don't mind just skipping the validation in download_zip_package().

---

seems like we always need to check conf.get_signature_validation_enabled() and openssl_version_supported_for_signature_validation(). can you create a method that calls both and use that as a single check?